From ac979de2d67c998203f016dd24c7f562b4372fce Mon Sep 17 00:00:00 2001
From: Rahul Akolkar <rahul@apache.org>
Date: Mon, 6 Jun 2011 22:47:09 +0000
Subject: [PATCH] SCXML-161 Transition leaving a child state of parallel
 incorrect. Fix by accounting for non-composite regions. Test case provided by
 Enrico Nardelli <nardelli at mat dot uniroma2 dot it>.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/scxml/trunk@1132819 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/scxml/SCXMLHelper.java |  2 +-
 .../commons/scxml/SCXMLExecutorTest.java      | 17 ++++++++-
 .../apache/commons/scxml/transitions-05.xml   | 38 +++++++++++++++++++
 3 files changed, 54 insertions(+), 3 deletions(-)
 create mode 100644 src/test/java/org/apache/commons/scxml/transitions-05.xml

diff --git a/src/main/java/org/apache/commons/scxml/SCXMLHelper.java b/src/main/java/org/apache/commons/scxml/SCXMLHelper.java
index 3b2411066..a5f8e0c0c 100644
--- a/src/main/java/org/apache/commons/scxml/SCXMLHelper.java
+++ b/src/main/java/org/apache/commons/scxml/SCXMLHelper.java
@@ -259,7 +259,7 @@ public static Set getStatesExited(final Transition t,
                     for (Iterator act = currentStates.iterator();
                             act.hasNext();) {
                         TransitionTarget a = (TransitionTarget) act.next();
-                        if (isDescendant(a, s)) {
+                        if (isDescendant(a, s) || a == s) {
                             //a is affected
                             boolean added = false;
                             added = allStates.add(a);
diff --git a/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java b/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java
index 568167b16..f993cd97d 100644
--- a/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java
+++ b/src/test/java/org/apache/commons/scxml/SCXMLExecutorTest.java
@@ -43,7 +43,7 @@ public SCXMLExecutorTest(String name) {
     // Test data
     private URL microwave01jsp, microwave02jsp, microwave01jexl,
         microwave02jexl, microwave03jexl, microwave04jexl, microwave05jexl, transitions01,
-        transitions02, transitions03, transitions04, prefix01, send01, send02;
+        transitions02, transitions03, transitions04, transitions05, prefix01, send01, send02;
     private SCXMLExecutor exec;
 
     /**
@@ -72,6 +72,8 @@ public void setUp() {
             getResource("org/apache/commons/scxml/transitions-03.xml");
         transitions04 = this.getClass().getClassLoader().
             getResource("org/apache/commons/scxml/transitions-04.xml");
+        transitions05 = this.getClass().getClassLoader().
+            getResource("org/apache/commons/scxml/transitions-05.xml");
         prefix01 = this.getClass().getClassLoader().
             getResource("org/apache/commons/scxml/prefix-01.xml");
         send01 = this.getClass().getClassLoader().
@@ -86,7 +88,7 @@ public void setUp() {
     public void tearDown() {
         microwave01jsp = microwave02jsp = microwave01jexl = microwave02jexl =
             microwave04jexl = microwave05jexl = transitions01 = transitions02 = transitions03 =
-            transitions04 = prefix01 = send01 = send02 = null;
+            transitions04 = transitions05 = prefix01 = send01 = send02 = null;
     }
 
     /**
@@ -234,6 +236,17 @@ public void testSCXMLExecutorTransitions04Sample() throws Exception {
             next()).getId());
     }
 
+    public void testSCXMLExecutorTransitions05Sample() throws Exception {
+        SCXML scxml = SCXMLTestHelper.parse(transitions05);
+        assertNotNull(scxml);
+        exec = SCXMLTestHelper.getExecutor(scxml);
+        assertNotNull(exec);
+        SCXMLTestHelper.assertPostTriggerStates(exec, "start", new String[]{"one", "two"});
+        SCXMLTestHelper.assertPostTriggerState(exec, "onetwo_three", "three");
+        SCXMLTestHelper.assertPostTriggerStates(exec, "three_one", new String[]{"one", "two"});
+        SCXMLTestHelper.assertPostTriggerState(exec, "two_four", "four");
+    }
+
     public void testSend01Sample() throws Exception {
         exec = SCXMLTestHelper.getExecutor(send01);
         assertNotNull(exec);
diff --git a/src/test/java/org/apache/commons/scxml/transitions-05.xml b/src/test/java/org/apache/commons/scxml/transitions-05.xml
new file mode 100644
index 000000000..7400359d8
--- /dev/null
+++ b/src/test/java/org/apache/commons/scxml/transitions-05.xml
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+-->
+<scxml xmlns="http://www.w3.org/2005/07/scxml" version="1.0" initial="init">
+       <state id="init">
+               <transition event="start" target="onetwo" />
+       </state>
+       <parallel id="onetwo">
+               <transition event="onetwo_three" target="three" />
+               <state id="one">
+               </state>
+               <state id="two">
+                       <transition event="two_four" target="four" />
+               </state>
+       </parallel>
+       <state id="three">
+               <transition event="three_one" target="one" />
+               <transition event="three_four" target="four" />
+       </state>
+       <state id="four">
+               <transition event="four_onetwo" target="onetwo" />
+               <transition event="four_three" target="three" />
+       </state>
+</scxml>
